### PR TITLE
Extract javaagent-extension-api from tooling & spi

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Implementation of (a part of) the module split discussed here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/2779

Sorry for huge PR, but I couldn't really split it further - fortunately most of it are renames and import changes.

All classes used to implement a javaagent instrumentation/extension were extracted to the `javaagent-extension-api` module:
Screenshot 2021-04-27 at 17 09 27
All javaagent instrumentations are now using this module and do not depend on `javaagent-tooling` directly; everything that was left in `javaagent-tooling` is an internal javaagent component that's not supposed to be exposed or used by instrumentation developers.

I've introduced the `AgentExtension` interface which `InstrumentationModule` now implements; it also replaces `ByteBuddyAgentCustomizer` (which kind of resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1626). The `AgentExtension` serves as a base for any agent modification/enrichment, including instrumentation.

All interesting changes are in `javaagent-tooling` and `javaagent-extension-api`, almost everything else is just import renames.

### CODE REVIEW

**Comment by mateuszrzeszutek** [comment]
Ouch, this PR has so many changed files that it's almost impossible to review in GH... I'll have to figure out how to split it (at least split it to several commits) after all.

**Comment by trask** [comment]
> Ouch, this PR has so many changed files that it's almost impossible to review in GH... I'll have to figure out how to split it (at least split it to several commits) after all.

It's rendering for me

**Comment by trask** [comment]
> > Ouch, this PR has so many changed files that it's almost impossible to review in GH... I'll have to figure out how to split it (at least split it to several commits) after all.
> 
> It's rendering for me

I realized that's not what you said, ya, it's a lot of files, how about just a separate commit in this PR for the stuff under `javaagent-tooling` and `javaagent-extension-api`?

**Comment by mateuszrzeszutek** [comment]
> I realized that's not what you said, ya, it's a lot of files, how about just a separate commit in this PR for the stuff under javaagent-tooling and javaagent-extension-api?

I'll try to separate all renames/imports into a separate commit tomorrow, that should make reviewing the important stuff much easier.

**Comment by mateuszrzeszutek** [comment]
Okay, I've managed to split this PR into 2 commits:
* the first one contains (mostly) extracting the javaagent-extension-api module, almost no renames/moving between packages;
* the second one is moving `InstrumentationModule`, muzzle and type matchers and is pretty much just renames.

**Review by trask — APPROVED** [review]
thanks for tackling this!!

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentExtensionTooling.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Map;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.pool.TypePool;
+
+/**
+ * This interface contains methods for accessing or creating internal javaagent components that can
+ * be used to implement an {@link AgentExtension}.
+ */
+public interface AgentExtensionTooling {
```
[the remark this answers is not in the document]
Comment (iNikem) [reply javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentExtensionTooling.java:21]: but why is this public?

**Comment by iNikem** [comment]
I took only a quick look so far and my main concern is the number of public classes. E.g. do we really expect now that users will provide their own implementations of `InstrumentationContextProvider`? I understand that this is needed for a split between `javaagent-extension-api` and `javaagent-tooling` as @trask noted. But extension module uses them, I think, only inside `io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule#extend`. Should the body of this method actually live in tooling module?

Also, how exactly is this interface specific for instrumentation context?

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentExtensionTooling.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentExtensionTooling.java:21]: > it's needed to break the dependency from `javaagent-extension-api` on `javaagent-tooling`
>
> just wanted to check/confirm my understanding...

Yep, it's exactly that. This interface contains everything that we need from tooling/internals in instrumentations.

> but why is this public?

The whole interface? It's a parameter of one `AgentExtension` method which itself is public, and I suppose it does contain some useful things (like access to classloaders) for extension developers.

**Comment by mateuszrzeszutek** [comment]
> E.g. do we really expect now that users will provide their own implementations of `InstrumentationContextProvider`?

No, this interface is supposed to hide the actual implementation, not be implemented.

> But extension module uses them, I think, only inside `io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule#extend`. Should the body of this method actually live in tooling module?

That's an interesting idea - we would certainly be able to hide some more things this way. I'll think about how to do it nicely.

> Also, how exactly is this interface specific for instrumentation context?

With method names like `additionalInstrumentation()` - probably not at all? 😂 
More reasons to hide it

**Comment by mateuszrzeszutek** [comment]
> `io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule#extend` -  should the body of this method actually live in tooling module?

Done! The `AgentExtensionTooling` interface got considerably smaller, now I'm starting to think that maybe I could get rid of it completely... Please review the current implementation of `InstrumentationModule#extend()` and let me know what you think - I could use a similar trick to hide those muzzle deps.

**Comment by iNikem** [comment]
@mateuszrzeszutek What if `InstrumentationModule` should NOT be a `AgentExtension`? What if there should instead be `InstrumentationLoader implements AgentExtension`? And its `extend` method loads all `InstrumentationModule`, which provide all necessary information to apply each module to `AgentBuilder`. `InstrumentationLoader` will live `javaagent-tooling` module then and will take `AgentExtensionTooling` with it.

**Comment by mateuszrzeszutek** [comment]
> What if there should instead be `InstrumentationLoader implements AgentExtension`? And its `extend` method loads all `InstrumentationModule`, which provide all necessary information to apply each module to `AgentBuilder`. `InstrumentationLoader` will live `javaagent-tooling` module then and will take `AgentExtensionTooling` with it.

Hmm, interesting. If we move the `ReferenceMatcher` to the tooling module (and it's possible if we change the return type of `getMuzzleReferenceMatcher()` to `Reference[]`) then we could get rid of the `AgentExtensionTooling` class.
I kind of liked the fact that we had one SPI less, but that's not such a big deal after all - and we can simplify `javaagent-extension-api` so that certainly makes it worth it.

Do you mind if I implement it in another PR? This one is both hard to maintain (conflicts...) and hard to review...

**Comment by mateuszrzeszutek** [comment]
Okay, I moved all muzzle classes except `Reference` back to tooling and got rid of the `AgentExtensionTooling` interface. `InstrumentationModule` still implements `AgentExtension`, but now the change that @iNikem proposed should be really easy to implement.

**Comment by mateuszrzeszutek** [comment]
And now the package structure looks like that:
Screenshot 2021-05-06 at 16 02 50

**Comment by trask** [comment]
> I moved all muzzle classes except Reference back to tooling and got rid of the AgentExtensionTooling interface

👍👍👍

> Do you mind if I implement it in another PR? This one is both hard to maintain (conflicts...) and hard to review...

Yes, let's merge this! @iNikem

**Review by anuraaga — APPROVED** [review]

**Comment by iNikem** [comment]
> > I moved all muzzle classes except Reference back to tooling and got rid of the AgentExtensionTooling interface
> 
> 👍👍👍
> 
> > Do you mind if I implement it in another PR? This one is both hard to maintain (conflicts...) and hard to review...
> 
> Yes, let's merge this! @iNikem

I will trust you on that, @trask

**Pull request merged**